### PR TITLE
Build tweaks and documentation

### PR DESCRIPTION
**Pull request by qc00**
Partially addresses #9

### CODE REVIEW

File: docs/mkdocs/docs/technical/contributing.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+You can build the same image locally by running that script.
+We're looking into hosting the image in a public location.
+
+Windows
+-------
+We recommend using Visual Studio 2022 (or later) to install the compiler (MSVC v142 or newer) and tools
+(Windows SDK, CMake, Python).
+
+The Python that comes with Visual Studio is sufficient for creating release builds, but for debug builds, you will have
+to separately download from [Python.org](https://www.python.org/downloads/windows/).
```
Comment (poodlewars) [inline docs/mkdocs/docs/technical/contributing.md:37]: add a note to download a debug version of Python?

File: docs/mkdocs/docs/technical/contributing.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
+Setting up your OS dependencies
+===============================
+
+Linux
+-----
+The codebase and build system can work with any reasonably recent Linux distribution that ships at least
+GCC 8 (10+ recommended) and CMake 3.12 (3.18+ recommended).
+
+A development install of Python 3.6+ (with `libpython.a` or `.so` and full headers) is also necessary.
+(See [pybind11 configuration](#pybind11-configuration))
```
Comment (poodlewars) [inline docs/mkdocs/docs/technical/contributing.md:10]: We shouldn't have the brackets after the full stop

File: docs/mkdocs/docs/technical/contributing.md (hunk 1 above)
Comment (qc00) [reply docs/mkdocs/docs/technical/contributing.md:37]: I did mention that I'll leave this section mostly out for you to contribute :P

File: docs/mkdocs/docs/technical/contributing.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+}
+```
+</details>
+
+## vcpkg caching
+We use [vcpkg](https://vcpkg.io/) to manage the C++ dependencies.
+
+The CMakePresets we provided use the `cpp/vcpkg` directory as the toolchain, so
+it's not necessary to separately download and install vcpkg.
+(You can, however, maintain a separate vcpkg install and override the CMake presets.)
```
Comment (poodlewars) [inline docs/mkdocs/docs/technical/contributing.md:101]: Should not have brackets after the full stop

File: docs/mkdocs/docs/technical/contributing.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+We augmented pybind11's [Python discovery](https://pybind11.readthedocs.io/en/stable/compiling.html#building-with-cmake)
+with our own [PythonUtils](https://github.com/man-group/ArcticDB/blob/master/cpp/CMake/PythonUtils.cmake) to improve
+diagnostics.
+Please pay attention to warning messages from `PythonUtils.cmake` in the CMake output which highlights any configuration
+issues with Python.
+
+We recommend:
+
+* Using the latest [manylinux](#manylinux-docker-image) image for Linux development where possible;
+* Or, getting the latest CMake (at least 3.18) and make sure it's the first on your PATH
```
Comment (poodlewars) [inline docs/mkdocs/docs/technical/contributing.md:127]: don't we require 3.21+?

File: docs/mkdocs/docs/technical/contributing.md [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+* [Asset caching](https://learn.microsoft.com/en-us/vcpkg/users/assetcaching)
+
+For example, you can use a network share as the caching folder and put the caching configuration environment
+variables into the CMake presets.
+
+Please also note compiling the dependencies uses a lot of disk space.
+Once CMake configuration is done and all the dependencies saved into the binary cache,
+you can remove the `cpp\vcpkg\buildtrees` folder.
+
+## pybind11 configuration
```
Comment (poodlewars) [inline docs/mkdocs/docs/technical/contributing.md:117]: I don't think we should write so much about this. We should just describe how to flip between Python versions with `Python_ROOT_DIR` and have that be our first-class approach.

File: python/arcticdb/util/hypothesis.py [hunk 6]
```diff
@@ -43,6 +43,14 @@ def test_something(function_scope_fixture):
     if not _function_scoped_fixture:
         return fun
 
+    existing = getattr(fun, "_hypothesis_internal_use_settings", None)
```
Comment (poodlewars) [inline python/arcticdb/util/hypothesis.py:46]: This is already in a different PR, no?

**Review by poodlewars — CHANGES_REQUESTED** [review]

File: docs/mkdocs/docs/technical/contributing.md [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+      "name": "alt-vcpkg-debug:py3.10",
+      "configurePreset": "alt-vcpkg-debug:py3.10",
+      "inheritConfigureEnvironment": true
+    }
+  ]
+}
+```
+</details>
+
+## vcpkg caching
```
[the remark this answers is not in the document]
Comment (mehertz) [reply docs/mkdocs/docs/technical/contributing.md:96]: I think we might have overdocumented here :) There's nothing specific to Arctic here is there?

File: docs/mkdocs/docs/technical/contributing.md [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+## pybind11 configuration
+We augmented pybind11's [Python discovery](https://pybind11.readthedocs.io/en/stable/compiling.html#building-with-cmake)
+with our own [PythonUtils](https://github.com/man-group/ArcticDB/blob/master/cpp/CMake/PythonUtils.cmake) to improve
+diagnostics.
+Please pay attention to warning messages from `PythonUtils.cmake` in the CMake output which highlights any configuration
+issues with Python.
+
+We recommend:
+
+* Using the latest [manylinux](#manylinux-docker-image) image for Linux development where possible;
```
Comment (mehertz) [inline docs/mkdocs/docs/technical/contributing.md:126]: Then this should be front and center. In fact maybe entirety of the instructions should be about building using this image.

File: docs/mkdocs/docs/technical/contributing.md (hunk 4 above)
Comment (qc00) [reply docs/mkdocs/docs/technical/contributing.md:127]: 3.21 is to support version 3 of the CMakePresets.
3.18 is sufficient to find the `Development.Module`.

File: docs/mkdocs/docs/technical/contributing.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (qc00) [reply docs/mkdocs/docs/technical/contributing.md:96]: I'll simplify

File: docs/mkdocs/docs/technical/contributing.md (hunk 5 above)
Comment (qc00) [reply docs/mkdocs/docs/technical/contributing.md:117]: It is the part that is the most poorly documented by third-party tools and most likely to go wrong.

Before I figure out what exactly is happening, I found probably 30 different ways to do it wrong. Including compiling against Python 3.10 headers but linking to 3.6 library (the most confusing errors).

I doubt a typical develop will do it right the first time without guidance.

File: python/arcticdb/util/hypothesis.py (hunk 6 above)
Comment (qc00) [reply python/arcticdb/util/hypothesis.py:46]: Rebased out

File: docs/mkdocs/docs/technical/contributing.md (hunk 8 above)
Comment (qc00) [reply docs/mkdocs/docs/technical/contributing.md:126]: What about users of other operating systems?

Comment (qc00) [reply docs/mkdocs/docs/technical/contributing.md:126]: I'll add some content, but it'll be difficult to harmonise with the development experience of other operating systems...

File: docs/mkdocs/docs/technical/contributing.md (hunk 5 above)
Comment (poodlewars) [reply docs/mkdocs/docs/technical/contributing.md:117]: Or really good cmakepresets with a well documented single thing to tweak?

**Review by poodlewars — APPROVED** [review]

**Pull request merged**